### DISCUSSION TITLE
User Guide Restructuring

### DISCUSSION THREAD

**Discussion by riedgar-ms** [fairlearn/fairlearn#658] [General]
I'd like to propose restructuring the 'mitigation' section of the user guide. It's already quite a large file, and we know that it's far from complete.

I suggest that we create a `mitigations` directory along these lines:
```
mitigations/
  index.rst
  postprocessing.rst
  preprocessing.rst
  reductions.rst
```
The `index.rst` file would contain the introductory part of the existing `mitigation.rst` file, with the other two then taking the content from the appropriate sections of that file. Within each file, each section could be promoted by one level.

For writing these files, we should try to make the deeper, more mathematical sections skippable. Each could start by saying something like "In this section, we will prove the result `1+1=2`; if you are prepared to take this on trust, you can skip to the next section."

Tagging @hildeweerts  and @romanlutz  who I know have opinions about some of these things.

**Comment by hildeweerts** [comment]
Sounds good to me!

**Reply by romanlutz** [reply]
Same. I would prefer having the order be
- preprocessing
- reductions
- postprocessing
for obvious reasons.

Regarding adding more math and making it "skippable" I brought that up in the community call chat. For completeness I'll add the comments from that chat:
- @adrinjalali : "I'd say if it's not in the API doc, and it's in the user guide, as long as we have a nice ToC on the side, we can as much as one wants"
- @MiroDudik : "My expectation is that the user guide should answer the question 'what do i need to know to use this'. For example a user guide for a python dictionary does not need to talk about the data structures it uses. @adrinjalali --agreed. Good table of contents is key..."

I think this is very much in line with what @riedgar-ms wrote, although you should feel free to correct me if I misinterpreted something :-)

**Reply by hildeweerts** [reply]
Do we at some point want to add more ‘in-processing’ methods in addition to reductions? Otherwise it might make sense to call the folder inprocessing instead of reductions.

**Reply by adrinjalali** [reply]
I'd also be happy to renaming `reduction` to `inprocessing`. Reduction doesn't seem an intuitive term for newcommers.

**Reply by riedgar-ms** [reply]
PR to kick this off:
https://github.com/fairlearn/fairlearn/pull/661

**Reply by MiroDudik** [reply]
I was considering `inprocessing`, but in the end I was convinced against it: see [comment1](https://github.com/fairlearn/fairlearn/issues/466#issuecomment-650186630) and [comment2](https://github.com/fairlearn/fairlearn/issues/466#issuecomment-650564824).

I take @adrinjalali 's point about clarity re. `reductions`, so we should make things clear in the user guide. I just don't like the term `inprocessing` very much (it suggests data transformation rather than training, and it is not as widely accepted as I had initially thought).

Can we keep preprocessing / postprocessing / reductions (or in a different order if that's preferred) for now? We could discuss `inprocessing` at some point in future. If people feel strongly about it, we can discuss in the next call?

**Reply by adrinjalali** [reply]
> it suggests data transformation rather than training, and it is not as widely accepted as I had initially thought

at least to me it means more of a loss function manipulation than anything. But it could also mean data transformation. We could also just call it `models` or something.

The more we wait, the harder it gets to change a module's name.

**Reply by hildeweerts** [reply]
I personally strongly dislike `reductions` as a separate module. I understand why it made sense at the time, but if we want to grow Fairlearn into a general fair ML project - that doesn't look like it's an implementation of a single paper anymore - it doesn't make sense to me in the long run. 

Full disclosure: from the current user guide and API it was very unclear to me at first why the approach was called reductions and how it should be used*. I am also not a fan of the term inprocessing, but I'd say that lack of clarity and wide acceptance are even more of a concern for `reductions` than for `inprocessing`. 

@adrinjalali raises a very valid point re. naming. I think this is something we need to consider this carefully.

I think the distinction between pre-processing and other approaches is usually very clear, but the line between inprocessing and postprocessing methods isn't. For example, from what I understand, you could also classify reductions as a post-processing method, because it doesn't directly add a fairness constraint to the ML algorithm's loss function, but acts more like a wrapper of an existing classifier.

Instead of `inprocessing` and `postprocessing`, we could also distinguish between `wrappers` and `models`, where wrappers take as input a user-supplied estimator, and models are fairness-aware versions of specific learning algorithms. If `models` is too general, we can also go full scikit-learn with `classification`, `regression`, `linear_model`, etc.

*From a user perspective, I also don't like that the fairness constraints need to be defined as separate objects and that it is not directly compatible with the `metrics` module, but let's leave that to a separate discussion.

**Reply by romanlutz** [reply]
I agree with most of the comments above, specifically
- not waiting too long to settle on something
- `reductions` isn't a great (top-level) name since most people don't intuitively understand what that means without reading more about it.
- `inprocessing` isn't ideal either, but certainly more intuitive and opens the door to add more "inprocessing" techniques. The reductions directory would simply move "one level deeper".

Additionally:
- I'm not sure I like `models`. The result of `postprocessing` is also a model. Wrapping is similarly nebulous to me. Is it wrapping only if I pass in an estimator (like in `ExponentiatedGradient`)? Or is it also wrapping if I only wrap SVMs internally and the user doesn't pass them explicitly? I think what we mean by wrapper here is meta-estimator. Perhaps that should just be the module name?
- `classification` and `regression` are currently clubbed together since `ExponentiatedGradient` and `GridSearch` can handle both.

Re: fairness constraint objects --> the original thinking there was to be able to combine them, although nobody has had time for that. 
For combining them we'd need something more elaborate than just simple strings like "demographic_parity". Imagine specifying 
- DemographicParity & Error Rate parity, or 
- DP on one sensitive feature, and DP on another sensitive feature, and DP on the intersections (which is now a much stronger DP constraint than either one of them individually)  
I agree on taking a separate discussion for that, though :-)

**Reply by MiroDudik** [reply]
First off: I think that it might be worth to discuss this on a call! Also, I think that the topic has shifted to the module structure from the documentation structure (which is a good thing to discuss and definitely a good thing to settle on early).

**Module structure**. When we came up with the initial module names (`postprocessing`, `reductions`, `metrics`, `widget`), I was very much inspired by the scikit-learn's shallow approach to modules, meaning many modules of various sizes, but no submodules. So `reduction` module was never intended to contain anything beyond `ExponentiatedGradient`, `GridSearch` and various `Moment` classes. It would be similar to some smaller scikit-learn modules like [sklearn.mixture](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.mixture) or [sklearn.naive_bayes](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.naive_bayes). I think those are fine modules. If we keep `reductions` module restricted to the implementation of algorithms from the two papers that both have the word "reduction" in their title, do you still think we should change the name of the module? Of course, we should definitely do a better job explaining what's in it...

As we add new algorithms etc., we can figure out whether they make sense to be added into any of the existing modules--for example, there are postprocessing approaches for regression that do not seek to optimize a threshold--or whether we need to create a new module.

With this conception of modules, we shouldn't be creating large modules that beg for splitting into submodules. This is an argument against something very broad like `inprocessing` or `models`. But we could have something like `linear_model`, which would then house mitigation algorithms that are specifically tailored for mitigation of linear models.

So, to frame the discussion about the module structure, we need to decide whether we want to go beyond one level of hierarchy. I personally don't think we should do it, because it becomes clunky to use and it's easy to get the hierarchy wrong. But we should discuss.

**Circling back to the documentation structure**. The user guide can of course be structured differently. We already have the "assessment" part which covers `fairlearn.metrics` and `fairlearn.widget` (soon to be gone), and "mitigation" part which covers other modules. We can structure the documentation however it makes sense. I would structure mitigation section as:

* Preprocessing approaches
* Direct optimization approaches
  (Right now we have none, but if there's movement on #466, it would go into this section.)
* Postprocessing approaches
* Optimization via reductions

Or we could possibly combine the two optimization sections into one. This is compatible with the terminology in [fairmlbook](https://fairmlbook.org/classification.html) -- search on page for "interlude".

Thoughts?

**Reply by adrinjalali** [reply]
I think @MiroDudik 's point on shallow module structure is very valid, and I like it.

In terms of document structure, not sure if it's in the scope of this thread: what do folks think about having separate pages for each class, instead of having them all in one page?

**Reply by romanlutz** [reply]
I have to say the organization makes sense to me as well the way you explained it @MiroDudik . Is the fourth bucket really just about reductions, though, or could there be other meta-estimation techniques (which then could get their own modules)? [Regardless I don't think it takes away from your point]

To @adrinjalali 's point:
I actually never quite understood why people want separate pages since you can link to specific sections within a page. I much prefer being able to CTRL+F to whatever part of the page I want rather than having to navigate through the tree of pages. Somehow in my memory that was what scikit-learn does as well, until I checked and realized that it's not! I think my confusion stems from there being a LOT more content per class/method, e.g., [this page](https://scikit-learn.org/stable/modules/sgd.html) on SGD alone is almost as long as our entire mitigation user guide. That shouldn't be a good reason not to split it up into individual pages, though. Rather, it should be a hint to us to finally spend time writing proper user guides. I think this is very much in line with the original intended purpose of this thread, which is "user guide restructuring". I am in favor of individual pages if that's up for debate :-)
